cpp/ql/src/semmle/code/cpp/ir/internal/Instruction.qll

```diff
@@ -67,6 +67,31 @@ module InstructionSanity {
     not tag instanceof UnmodeledUseOperand
   }
 
+  /**
+   * Holds if `Phi` instruction `instr` has fewer than two operands.
+   */
+  query predicate missingPhiOperands(PhiInstruction instr, int predIndex, Location predLoc) {
+    exists(IRBlock pred |
+      pred = instr.getBlock().getAPredecessor() and
+      predLoc = pred.getLocation() and
+      predIndex = pred.getDisplayIndex() and
+      not exists(PhiOperand operand |
+        exists(instr.getOperand(operand)) and
+        operand.getPredecessorBlock() = pred
+      )
```

Comment:
This predicate doesn't implement what its QLDoc says. It checks to make sure phi instructions have operands for all of their predecessors. But why should that always hold? After `int x = 0; if (b) { x = 1; } else { f(); }` I'd expect a phi node with two operands, one for each `x =` but none for the `f()` block.

Reply:
`PhiOperand.getPredecessorBlock()` specifies the immediate predecessor block from which the value flowed, not the block that contains the definition. This is important for anything trying to ignore unreachable edges, or anything vaguely path sensitive, because we need to know which definitions flowed in on which edge, even if the same definition came via multiple edges.

Reply:
I see. Then I'm happy with the check but not with its QLDoc. Could you add a separate check that any block with a phi node has at least two predecessors?

Reply:
Fixed the comment and added `unnecessaryPhiInstruction`.